tests/values.go

```diff
@@ -60,14 +60,14 @@ type Chart struct {
 //
 // TODO: generate these structs from a values.schema.json
 type CoderValues struct {
-	Certs    *CertsValues    `json:"certs" yaml:"certs"`
-	Coderd   *CoderdValues   `json:"coderd" yaml:"coderd"`
-	Envbox   *EnvboxValues   `json:"envbox" yaml:"envbox"`
-	Ingress  *IngressValues  `json:"ingress" yaml:"ingress"`
-	Logging  *LoggingValues  `json:"logging" yaml:"logging"`
-	Metrics  *MetricsValues  `json:"metrics" yaml:"metrics"`
-	Postgres *PostgresValues `json:"postgres" yaml:"postgres"`
-	Services *ServicesValues `json:"services" yaml:"services"`
+	Certs    *CertsValues    `json:"certs,omitempty" yaml:"certs,omitempty"`
```

Comment:
omitempty doesn't really matter here, since these will always be set, based on the defaults in the Chart values.yaml, right?

Reply:
When I tested it, it seemed like since it was rendering them all as null, it was overriding the defaults in the Chart `values.yaml`.

Reply:
So a bit of context on the whole CoderValues struct and why it exists... I wanted code completion, because otherwise the Helm API would result in a nested literal situation like:

```go
values := map[string]interface{}{
    "coderd": map[string]interface{}{
        "securityContext": map[string]interface{}{
            "runAsNonRoot": true,
            "seccompProfile": map[string]interface{}{
                "type": "RuntimeDefault",
            },
        },
    },
}
```

This seems kinda ugly and we also lose code completion, as well as validation. I then learned that the reason Helm does it this way (I guess) is that this allows people to distinguish between values being nil (key exists and set to null) and undefined (key not defined).

So the CoderValues always returns the full values defined in the original chart, then the caller needs to copy and modify it, and pass it back. Otherwise, if you just pass a CoderValues struct, then there's no way to distinguish between the unset/undefined (in which case, we want to apply the default) or null (in which case, we want to override the default) cases.

So these things should never be null, otherwise they get passed into Helm as null. If you use omitempty, then there's no way to explicitly set things to null to override defaults either. Either way, we're guessing between one of the cases (unset/undefined, override to null) and in some of the cases our guess will be incorrect (and unfixable) :frowning:

Reply:
Would it be better if I just embedded a yaml string then?